### PR TITLE
[avr] Support avrdude fusebits via options

### PR DESCRIPTION
**Pull request by TomSaw**
**This has no priority!**
But when it flows, it flows. 🏄

I'm not 100% intimate with the inner workings of lbuild. My implementation of fusebit-options for AVRs is rather copy-catted than wisely coded. It works pretty good but again let me know, if i missed a thing, or two. 

- validate_fuse may be over-engineered
- Is :build:avrdude the right scope for the fusebits? They do not depend on the programmer itself!

Edit: in src/modm/board/arduino_nano/board.xml i've set the default fusebits for arduino-nano, so you can lbuild one of the nano-examples and quickly play around.

### CODE REVIEW

File: tools/build_script_generator/common.py [hunk 1]
```diff
@@ -154,11 +160,20 @@ def common_avrdude_options(env):
         option_baudrate = env.collector_values(":build:default.avrdude.baudrate", option_baudrate)[0]
     if not option_options:
         option_options = env.collector_values(":build:default.avrdude.options", option_options)[0]
+    if not option_hfuse:
+        option_options = env.collector_values(":build:default.avrdude.hfuse", option_hfuse)[0]
+    if not option_lfuse:
+        option_options = env.collector_values(":build:default.avrdude.lfuse", option_lfuse)[0]
+    if not option_efuse:
+        option_options = env.collector_values(":build:default.avrdude.efuse", option_efuse)[0]
```
Comment (salkinium) [inline tools/build_script_generator/common.py]: This only makes sense if there are some default collectors for these too.

**Comment by salkinium** [comment]
Ah, wait a second. The idea was to program the fuses via the ELF file mechanism, see this: https://www.nongnu.org/avr-libc/user-manual/group__avr__fuse.html
That's why this isn't implemented yet.

I guess we should document this somewhere… 😬😳

**Comment by TomSaw** [comment]
Ohhh No 😓

**Comment by salkinium** [comment]
The issue is that avrdude doesn't support a config file, like `openocd.cfg`, so I didn't want to pipe all options through lbuild, since that just bloats it all up, only the absolutely necessary options.

The question is, whether you can create a `modm_weak FUSES = { ...defaults... };` in the `board.cpp` file, which can be overwritten by an application?

**Comment by TomSaw** [comment]
There are various fractals named 'fuses' in the python-files.  I thought fusebits have not been re-implemented after a bigger refactorisation.

Do you need thes fractals for the new implementation?

**Comment by salkinium** [comment]
> Do you need thes fractals for the new implementation?

"fractals for the new implementation"? 🤨

The FUSES struct should already work with the code today. I remember testing it in hardware.

**Comment by TomSaw** [comment]
Ok, this goes beyond my skillz - would have to learn more background.

- I can put the fuses into a struct in my local main.cpp?
- How does 'scons program-fuses' recognize them?

**Comment by salkinium** [comment]
> I can put the fuses into a struct in my local main.cpp?

Yes, like this:
```
FUSES = 
{
    0xFF, // .low
    0xDA, // .high
    0xFD, // .extended
};
```

> How does 'scons program-fuses' recognize them?

The `FUSES` is a macro expanding to `__fuse_t __fuse __attribute__((section (".fuse"))`, which puts the `__fuse` object into the linkersection `.fuse`, which is not part of the application binary, it is only for avrdude, which can find and read the values from the ELF file. We just need to tell avrdude, which fuses we want programmed (all by default for scons).

**Comment by TomSaw** [comment]
Yea got this  :).. you have even the #Makros:

FUSES = 
{
    .low = LFUSE_DEFAULT,
    .high = HFUSE_DEFAULT,
    .extended = EFUSE_DEFAULT,
};


but 'scons program-fuses' still produces:
`avrdude -V -p m328p -c avrisp2 -P usb -b 115200 -U flash:w:build/release/Sorter.elf -U hfuse:w:build/release/Sorter.elf -U lfuse:w:build/release/Sorter.elf -U efuse:w:build/release/Sorter.elf`

**Comment by salkinium** [comment]
> but 'scons program-fuses' still produces:

Yes, that's correct. The format is `section name : operation : source` with the source being the ELF file.

**Comment by TomSaw** [comment]
Cool this sound interesting. I'll check it out.

**Comment by salkinium** [comment]
See https://www.nongnu.org/avrdude/user-manual/avrdude_4.html#Option-Descriptions `-U memtype:op:filename[:format]`

**Comment by salkinium** [comment]
The interesting question is, if you can define a weak default struct in the BSP `board.cpp` and *overwrite* it in the application:

in `modm/src/board/board.cpp` (`modm_weak` = weak linkage):
```c
modm_weak FUSES = 
{
    LFUSE_DEFAULT, // .low
    HFUSE_DEFAULT, // .high
    EFUSE_DEFAULT, // .extended
};
```
(or non-default values if required for the specific board).

Then in `main.cpp` to overwrite these (no `modm_weak`, ie. strong linkage):
```c
FUSES = 
{
    0xFF, // .low
    0xDA, // .high
    0xFD, // .extended
};
```

**Comment by TomSaw** [comment]
Ok, there's plenty of stuff i need to read + understand before digging deeper! Gimme some ⏰ and i'll be back in the conversation.

**Comment by salkinium** [comment]
> if you can define a weak default struct in the BSP board.cpp and overwrite it in the application

Yes, just tested that and it works as expected. However, this is completely undocumented right now, I'm adding docs for `scons program-fuses`. I guess that would be the right place, since people will hopefully check the build system for the right commands.

> Gimme some ⏰ and i'll be back in the conversation.

Sure, no stress, happy weekend!

**Comment by TomSaw** [comment]
I'm longing for your docs.
Happy weekend back!

**Comment by TomSaw** [comment]
Ahhh 😏 Weekend. Project-Time!

**Comment by salkinium** [comment]
I'm closing this in favor of the ELF mechanism. Sorry for the lost work!

**Comment by TomSaw** [comment]
No prob. Learned some about lbuild. And FUSES and ELF.

For me is newland, so honestly, thanks for the clarification and good luck with the ELF.

I'll throw an eye on #163 as a compensation.

**Pull request closed without merging**